Replace all concatenated paths with Node's path.join

astorije opened the pull request:
My attempt at fixing #301.

Let me know if this solution is crappy or should be improved, I don't claim this is the one solution and I'm happy to get feedback. Also, I hope I didn't forget anything.

xPaw commented on src/userLog.js:
```diff
@@ -5,7 +5,7 @@ var Helper = require("./helper");
 
 module.exports.write = function(user, network, chan, msg) {
 	try {
```
Can we move this whole try/catch block into main `exports`, so that it's not invoked on each `write`? Alternatively this can be moved all the way to `index` CLI where `config` is created.

xPaw commented on src/helper.js:
```diff
@@ -1,14 +1,23 @@
+
 function getConfig() {
-	return require(path.resolve(this.HOME) + "/config");
+	return require(path.resolve(this.CONFIG_PATH));
```
Remove `path.resolve` here.

xPaw commented on src/helper.js:
```diff
@@ -1,14 +1,23 @@
 var path = require("path");
 var os = require("os");
 
+var HOME = path.join(process.env.HOME || process.env.USERPROFILE, ".lounge");
```
Add `path.resolve` here.

xPaw commented on src/command-line/index.js:
```diff
@@ -15,12 +15,12 @@ if (program.home) {
 	fs.writeFileSync(
 		config,
-		fs.readFileSync(__dirname + "/../../defaults/config.js")
+		fs.readFileSync(require("path").join(__dirname, "..", "..", "defaults", "config.js"))
```
I think `path.resolve` is needed here.

xPaw commented on src/helper.js:
```diff
@@ -26,4 +35,3 @@ function expandHome(path) {
 
 	return path.replace(/^~($|\/|\\)/, home + "$1");
```
Add `path.resolve` after replacing.

xPaw commented on src/command-line/index.js:
```diff
@@ -15,12 +15,12 @@ if (program.home) {
 	Helper.HOME = program.home;
```
Add `path.resolve` here.

astorije commented:
@xPaw, I believe I have addressed all your comments.

However, this PR needs a lot of testing. I have done some, but definitely happy to have others test too. If done badly, this can create and use config in wrong directories, so people would believe their config was wiped.

astorije commented on src/command-line/index.js:
```diff
@@ -12,15 +13,21 @@ program.option("    --home <path>" , "home path");
 		config,
-		fs.readFileSync(__dirname + "/../../defaults/config.js")
+		fs.readFileSync(path.resolve(path.join(
+			__dirname,
```
[This](https://github.com/thelounge/lounge/blob/c8df0d68c02f30a9cb0c7c7fa93d93479654223b/src/plugins/irc-events/ctcp.js#L1) uses something else, I'm thinking the goal is the same. Should we stick with one or the other?

xPaw replied:
`__dirname` is correct.

xPaw commented on src/helper.js:
```diff
@@ -1,17 +1,26 @@
-	HOME: (process.env.HOME || process.env.USERPROFILE) + "/.lounge",
-	getConfig: getConfig,
+	HOME: HOME,
+	CONFIG_PATH: path.join(HOME, "config.js"),
```
We discussed this in IRC, but for sake of visibility: this breaks `--home` option.

xPaw commented:
Okay, I've fixed `--home`. It also correctly expands `~` if specified.

astorije commented:
Alright, I'm all set with this PR! (Removing `second review needed` because I made changes since @xPaw's additions).

@thelounge/maintainers, could you review and test this please? :-)

YaManicKill commented:
Review looks fine, but I'll set it up so I can test it tomorrow probably (quite busy tonight).

astorije commented:
@maxpoulin64, do you mind giving this a review as well?
